### ISSUE TITLE
Subscription returning 405 Method Not Allowed

### ISSUE THREAD

**Issue by MarkRuse** [EntityGraphQL/EntityGraphQL#347]
I have successfully implemented Query and Mutation within my solution but am struggling to get subscriptions to work - regardless of the request query (if it even exists or not) I am getting a 405 Method not Allowed.  I have tried to run the example codebase but am getting issues with the Antlr setup so instead I removed the direct project reference and included EntityGraphQl 5.1.1 in the project and getting the same error.  I think it may well be down to the program.cs wiring but am at a loss after many days of AI / web searching.   Interestingly, if I change my subscription code to an object that doesn't exist I get back the same 405 error so it looks like it isn't getting to the underlying code.

My Program.cs is as follows

```c#
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddControllers();

builder.Services.AddDbContext<TestContext>(options => options.UseSqlServer(builder.Configuration.GetConnectionString("DefaultConnection")).LogTo(Console.WriteLine, LogLevel.Trace));

builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

builder.Services.AddScoped<PersonService>();

builder.Services.AddGraphQLSchema<TestContext>(options =>
    {
        options.ConfigureSchema = GraphQLSchema.ConfigureSchema;
    })
    .AddGraphQLValidator();

var app = builder.Build();
app.UseWebSockets();

app.UseRouting();
app.MapGraphQL<TestContext>();
app.MapControllers();

//app.UseAuthorization();

//app.UseGraphQLWebSockets<TestContext>();

app.Run();
``` 

my person service is as follows

```c#
    public class PersonService
    {
        private readonly TestContext _context;
        private readonly Broadcaster<Person> broadcaster = new();

        public PersonService(TestContext context)
        {
            _context = context;
        }

        [GraphQLMutation("Mutate Person")]
        public Expression<Func<TestContext, Person>> PersonMutate(TestContext db,
            PersonInput args, IGraphQLValidator validator)
        {
            var person = new Person()
            {
                Id = null,
                Name = args.Name,
                Age = args.Age
            };

            db.Persons.Add(person);
            db.SaveChanges();

            broadcaster.OnNext(person);

            return (ctx) => ctx.Persons.First(p => p.Id == person.Id);
        }

        public IObservable<Person> Subscribe()
        {
            return broadcaster;
        }
    }
``` 

and my PersonSubscription is 
```c#
    public class PersonSubscription
    {
        [GraphQLSubscription("Example of a subscription")]
        public IObservable<Person> OnPersonUpdate(PersonService service)
        {
            return service.Subscribe();
        }
    }
``` 

My schema is as follows

```c#
        public static void ConfigureSchema(SchemaProvider<TestContext> schema)
        {
            schema.AddInputType<PersonInput>("PersonInput", "New person data")
                .AddAllFields();

            schema.Mutation().AddFrom<PersonService>();

            schema.Subscription().AddFrom<PersonSubscription>();

            File.WriteAllText("schema.graphql", schema.ToGraphQLSchemaString());
        }
``` 

When I call via Postman (using the GraphQL query type)

```typescript
subscription OnPersonUpdate {
    onPersonUpdate {
        id
        name
        age
    }
}

``` 
I get back this response
```javascript
Unexpected server response: 405
Handshake Details
Request Method: "GET"
Status Code: "405 Method Not Allowed"
Request Headers
Sec-WebSocket-Version: "13"
Sec-WebSocket-Key: "zNyXCTzKoT0QHn3NR7RqiQ=="
Connection: "Upgrade"
Upgrade: "websocket"
accept: "application/json"
content-type: "application/json"
user-agent: "PostmanClient/10.23.8 (AppId=a56be900-0128-4e7c-8fa5-ceb526eb4c90)"
Sec-WebSocket-Extensions: "permessage-deflate; client_max_window_bits"
Sec-WebSocket-Protocol: "graphql-transport-ws"
Host: "localhost:7129"
Response Headers
Content-Length: "0"
Date: "Mon, 04 Mar 2024 09:49:03 GMT"
Server: "Kestrel"
Allow: "POST"
``` 

Any help would be really appreciated as the library has been excellent so far

**Comment by lukemurray** [comment]
I assume you are running it with `app.UseGraphQLWebSockets<TestContext>();` not commented out?

Can you share more details on how you are configuring Postman? I cannot get Postman to make a valid web socket call to the GraphQL endpoint, it keeps sending a HTTP POST (It also looks like yours is sending a HTTP GET). I also can't find anything in the docs on Postman if it supports [this protocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md) which is the one we use.

**Comment by lukemurray** [comment]
OK I have reproduced what you get with Postman. It seems Postman sends a HTTP GET command. Which is not supported by EntityGraphQL for Subscriptions. We implement the GraphQL-WS protocol defined [here](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md) and supported by most client libraries.

I'll dig a little more into Postman, but I can;t find much about the protocol or methods it does for subscriptions

**Comment by lukemurray** [comment]
OK I think I figured this out.

The demo subscription project in the repo does work with Postman.

Note that by default EntityGraphQL registers 
- `POST` at `/graphql` for receiving queries and mutations
- `/subscriptions` where it expects it to be web sockets

In Postman you need to change the URL to `<your-address>/subscriptions` to have it work as that is the endpoint.

You can configure the endpoints. E.g. if you want both to be `/graphql`

```c#
app.UseGraphQLWebSockets<ChatContext>(path: "graphql");

app.UseEndpoints(endpoints =>
{
    endpoints.MapGraphQL<ChatContext>(); // defaults to "graphql"
});
```

Now in Postman you can use the same URL path for all query types. Not that if you do this you will need to make sure `app.UseGraphQLWebSockets` is before `MapGraphQL` otherwise `MapGraphQL` will return an error.

I hope this helps, let me know how you go. I'll aim to update the docs as well.

**Comment by MarkRuse** [comment]
Thanks Luke for taking a look at this - really appreciated.  First time implementing subscriptions so tried /graphql/subscriptions as saw something about this on the web but not just /subscriptions.  Will give it a try and confirm back.  Great news it is working in postman as well as that will most likely be the tool of choice for people implementing.

**Comment by lukemurray** [comment]
Yeh, Antlr sadly needs Java and often causes pain. I'll looking at removing the dependency in a 6.0 release.

I'll update the docs to mention the different path. Glad it is working.

**Comment by MarkRuse** [comment]
After solving my subscription issue, my plan was to keep separate eps for queries/mutations and subscriptions so didn't try the code to put them onto 1 ep.  Now after implementing a few it makes more sense to deal with all three types through one ep so I applied your code above to standardise on one ep

```c#
app.UseGraphQLWebSockets<TestContext>("/xxx");

app.UseEndpoints(endpoints =>
{
    endpoints.MapGraphQL<TestContext>(path: "/xxx"); // defaults to "graphql" but showing this as you could change both to whatever you want
});
```
I have included the / as the library throws an arg exception without it.  With the above code the subscriptions work perfectly however a call to a query will return  a 400 Bad Request.  I have tried this as well specifying "/graphql" as well with the same result.

Let me know if you want me to raise this as a new issue or  are happy to reopen this one
